pkg/controller/controller_instance.go

```diff
@@ -2021,3 +1830,216 @@ func shouldStartOrphanMitigation(statusCode int) bool {
 		statusCode == http.StatusRequestTimeout ||
 		is5XX
 }
+
+// serviceInstanceHasExistingBindings returns true if there are any existing
+// bindings associated with the given ServiceInstance.
+func (c *controller) serviceInstanceHasExistingBindings(instance *v1beta1.ServiceInstance) (bool, error) {
+	bindingLister := c.bindingLister.ServiceBindings(instance.Namespace)
+
+	selector := labels.NewSelector()
+	bindingList, err := bindingLister.List(selector)
+	if err != nil {
+		return false, err
+	}
+	for _, binding := range bindingList {
+		if instance.Name == binding.Spec.ServiceInstanceRef.Name {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// cloneServiceInstance clones the given ServiceInstance. This is so that the
+// reconciler may modify the cloned resource without accidentally mutating the
+// cache.
+func cloneServiceInstance(instance *v1beta1.ServiceInstance) (*v1beta1.ServiceInstance, error) {
+	clone, err := api.Scheme.DeepCopy(instance)
+	if err != nil {
+		return nil, err
+	}
+	return clone.(*v1beta1.ServiceInstance), nil
+}
+
+// updateServiceInstanceGracefulDeletionSuccess handles the logging and
+// updating of a ServiceInstance that has successfully finished graceful
+// deletion.
+func (c *controller) updateServiceInstanceGracefulDeletionSuccess(instance *v1beta1.ServiceInstance) error {
+	pcb := pretty.NewContextBuilder(pretty.ServiceInstance, instance.Namespace, instance.Name)
+
+	instance, err := cloneServiceInstance(instance)
+	if err != nil {
+		return fmt.Errorf(pcb.Messagef("error cloning: %v", err))
+	}
+
+	clearServiceInstanceCurrentOperation(instance)
```

Comment:
It doesn't really make sense to me to clear the current operation as part of a successful graceful delete, but I don't suppose it matters too much.

Reply:
Hmm... I suppose that's true; if the operation strictly represents broker communication, then it is more appropriate just being in the `updateServiceInstanceDeprovisionSuccess` function. I'll move it.